File: wjdrud2532_c++/programmers/43165.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

vector<int> v;
int tnum;
bool isVisit[21] = {0, };
int answer = 0;

void dfs(int depth, int sum)
{
    if(v.size() == depth)
    {
        if(tnum == sum)
            answer ++;
        
        //cout << sum << " ";

        return ;
    }
    
    // for(int i = depth; i < v.size(); i ++)
    // {
    //     if(isVisit[i] == false)
    //     {
            // isVisit[i] = true;
            
            dfs(depth + 1, sum + v[depth]);
            dfs(depth + 1, sum - v[depth]);
            
    //         isVisit[i] = false;
    //     }
    // }
    
    return;
}

int solution(vector<int> numbers, int target) {
    
    for(int i = 0; i < numbers.size(); i ++)
        v.push_back(numbers[i]);
    
    tnum = target;
    
    dfs(0, 0);
    
    return answer;
}

File: belif_python/programmers/43165.py
def dfs(n, numbers, target, depth, result):
    if n == depth:
        count = 0
        
        if result + numbers[depth - 1] == target:
            count += 1
            
        if result - numbers[depth - 1] == target:
            count += 1
        
        return count
    
    x = dfs(n, numbers, target, depth + 1, result + numbers[depth - 1])
    y = dfs(n, numbers, target, depth + 1, result - numbers[depth - 1])
    
    return x + y

def solution(numbers, target):
    return dfs(len(numbers), numbers, target, 1, 0)